Add KGO Dataplane documentation (#1384)
Kong/developer.konghq.com: app/operator/dataplanes/konnectextension.md

---
title: "KonnectExtension"

> description: Automatically register DataPlanes with {{ site.konnect_short_name }} by providing authentication credentials in KonnectExtension

Reviewer: ```suggestion
description: Automatically register Data Planes with {{ site.konnect_short_name }} by providing authentication credentials in KonnectExtension
```

content_type: reference
layout: reference
products:
  - operator
breadcrumbs:
  - /operator/
  - index: operator
    group: Gateway Deployment
  - index: operator
    group: Gateway Deployment
    section: Key Concepts

---

Kong Dataplane instances can be configured in Konnect by means of the [`KonnectExtension`](/gateway-operator/dataplane/reference/custom-resources/#konnectextension-1) resource. This resource can be used to provision isolated `DataPlane`s (Hybrid mode) or pairs of `ControlPlane`s  and `DataPlane`s ({{ site.kic_product_name }} mode).

The `KonnectExtension` resource can be referenced by `ControlPlane`s, `DataPlane`s, or `GatewayConfiguration`s from the extension point in their spec. Dedicated guides will guide you to through these kinds of setup.

> ## Konnect ControlPlane reference

Reviewer: ```suggestion
## {{site.konnect_short_name}} Control Plane reference
```

> `KonnectExtension` can be attached to Konnect `ControlPlane`s of type Hybrid or KIC. This reference can be performed in two different ways: via Konnect ID or via Kubernetes object reference to an in cluster `KonnectGatewayControlPlane`.

Reviewer: ```suggestion
`KonnectExtension` can be attached to {{site.konnect_short_name}} `ControlPlane`s of type Hybrid or KIC. This reference can be performed in two different ways: via {{site.konnect_short_name}} ID or via Kubernetes object reference to an in cluster `KonnectGatewayControlPlane`.
```

### Reference by Konnect ID

> The Konnect ControlPlane can be referenced by its ID, without having any `KonnectGatewayControlPlane` resource deployed in the cluster. The ControlPlane ID can be fetched by the Konnect UI, in the ControlPlane page. Whit this configuration, the `KonnectExtension` object requires to have the `konnect.configuration.authref` field set, as follows in the snippet below:

Reviewer: ```suggestion
The {{site.konnect_short_name}} Control Plane can be referenced by its ID, without having any `KonnectGatewayControlPlane` resource deployed in the cluster. The Control Plane ID can be fetched by the {{site.konnect_short_name}} UI, in the Control Plane page. Whit this configuration, the `KonnectExtension` object requires to have the `konnect.configuration.authref` field set, as follows in the snippet below:
```

### Manual certificate provisioning

In order to manually create and set up a certificate to be used for Konnect `DataPlane`s, you can perform type the following commands:

1. Generate a new certificate and key:

    ```bash
    openssl req -new -x509 -nodes -newkey rsa:2048 -subj "/CN=kongdp/C=US" -keyout ./tls.key -out ./tls.crt
    ```

1. Create a Kubernetes secret that contains the previously created certificate:

    ```bash
    kubectl create secret tls konnect-client-tls --cert=./tls.crt --key=./tls.key
    ```

1. Label the secret to tell KGO to reconcile it:

    ```bash
    kubectl label secret konnect-client-tls konghq.com/konnect-dp-cert=true
    ```

Once the secret containing your certificate has been created in the cluster, you can set up your `KonnectExtension` as follows:

```yaml
spec:
  clientAuth:
    certificateSecret:
      provisioning: Manual
        secretRef:
          name: konnect-client-tls # The name of the secret containing your certificate
```

### Automatic certificate provisioning

> Alternatively, you can leave the certificate provisioning and management to KGO, which will take care of creating a new certificate, write it into a Kubernetes `Secret` and manage the `Secret`'s lifecycle on behalf of you. To do so, you can configure a `KonnectExtension` as follows:

Reviewer: ```suggestion
Alternatively, you can leave the certificate provisioning and management to {{ site.operator_product_name }}, which will take care of creating a new certificate, write it into a Kubernetes `Secret` and manage the `Secret`'s lifecycle on behalf of you. To do so, you can configure a `KonnectExtension` as follows:
```

```yaml
spec:
  clientAuth:
    certificateSecret:
      provisioning: Automatic
```

or you can just leave the `spec.clientAuth` field empty, and the automatic provisioning will be used as the default value.

> ## Set DataPlane labels

Reviewer: ```suggestion
## Set Data Plane labels
```

> Multiple labels can be configured to the Konnect `DataPlane` via the following field:

Reviewer: ```suggestion
Multiple labels can be configured to the {{site.konnect_short_name}} `DataPlane` via the following field:
```
